src/Infrastructure.VS.UnitTests/VsVersionProviderTests.cs

```diff
@@ -0,0 +1,140 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using FluentAssertions;
+using Microsoft.VisualStudio.Setup.Configuration;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Moq;
+using SonarLint.VisualStudio.Infrastructure.VS.VsVersion;
+using SonarLint.VisualStudio.Integration;
+using SonarLint.VisualStudio.Integration.UnitTests;
+using IVsShell = Microsoft.VisualStudio.Shell.Interop.IVsShell;
+
+namespace SonarLint.VisualStudio.Infrastructure.VS.UnitTests
+{
+    [TestClass]
+    public class VsVersionProviderTests
+    {
+        [TestMethod]
+        public void MefCtor_CheckIsExported()
+        {
+            MefTestHelpers.CheckTypeCanBeImported<VsVersionProvider, IVsVersionProvider>(null, new[]
+            {
+                MefTestHelpers.CreateExport<SVsServiceProvider>(CreateServiceProvider(Mock.Of<IVsShell>())),
+                MefTestHelpers.CreateExport<ILogger>(Mock.Of<ILogger>())
+            });
+        }
+
+        [TestMethod]
+        public void TryGet_CalculatesVersion()
+        {
+            const string name = "Microsoft Visual Studio Enterprise 2019";
+            const string buildVersion = "16.9.30914.41";
+            const string displayVersion = "16.9.0 Preview 3.0";
+            const string installDirectory = "some directory";
+
+            var vsShell = CreateVsShell(installDirectory);
+            var setupConfig = CreateSetupConfiguration(installDirectory, name, buildVersion, displayVersion);
+
+            var testSubject = CreateTestSubject(vsShell.Object, setupConfig.Object);
+            var vsVersion = testSubject.TryGet();
+
+            vsVersion.Should().NotBeNull();
+            vsVersion.ProductName.Should().Be(name);
+            vsVersion.ProductVersion.Should().Be(buildVersion);
+            vsVersion.ProductDisplayVersion.Should().Be(displayVersion);
+        }
+
+        [TestMethod]
+        public void TryGet_VersionIsCached()
+        {
+            const string installDirectory = "some directory";
+
+            var vsShell = CreateVsShell(installDirectory);
+            var setupConfig = CreateSetupConfiguration(installDirectory);
+
+            var testSubject = CreateTestSubject(vsShell.Object, setupConfig.Object);
+            
+            var vsVersion1 = testSubject.TryGet();
+
+            vsShell.Invocations.Clear();
+            setupConfig.Invocations.Clear();
+
+            var vsVersion2 = testSubject.TryGet();
+
+            vsVersion1.Should().BeSameAs(vsVersion2);
+
+            vsShell.Invocations.Should().BeEmpty();
+            setupConfig.Invocations.Should().BeEmpty();
+        }
+
+        [TestMethod]
+        public void TryGet_FailureToGetVersion_Null()
```

Comment:
I guess this is meant to test that the exception handler is hit. However, it's quite fragile as it's implicitly dependant on both the current test setup code and the product code implementation. It would be possible to change either inadvertently so the method returns null without hitting the exception handler.

It would be better if it made an explicit check e.g. that part of the `FailedToCalculateVsVersion` message got logged.

Reply:
I started with that originally, but it's a logDebug, which relies on the environment parameters... https://github.com/SonarSource/sonarlint-visualstudio/blob/master/src/Core.UnitTests/LoggerExtensionsTests.cs#L62

I'll change it not to be null, and instead throw and exception, and verify that it was called.